src/DirectoryMonitor.swift

```diff
@@ -0,0 +1,272 @@
+        self.eventHandler = eventHandler
+
+        let callback: FileSystemEventStreamCallback = { (_, context, _, _, _, _) -> Void in
+            if let eventHandler = unsafeBitCast(context, to: DirectoryMonitor.self).eventHandler {
+                eventHandler()
+            }
+        }
+
+        // The directory monitor will listen to events that happen in both
+        // directions of each directory's hierarchy and coalesce events that
+        // happen within 2 seconds of each other.
+        guard let stream = FileSystemEventStream(callback: callback, context: &context, directories: directories, latency: 2, configuration: [.fileEvents, .watchRoot]) else {
+            throw DirectoryMonitorError.streamCouldNotBeCreated
+        }
+
+        // FSEvents has a two-pronged approach: schedule a stream (and record
+        // the changes) and start sending events. This granualarity is not
+        // desirable for the directory monitor, so they are combined here.
+        if !isStreamCurrentlyScheduled {
+            stream.schedule()
+            isStreamCurrentlyScheduled = true
+        }
+
+        if !isStreamCurrentlyAllowedToSendEvents {
+            do {
+                try stream.start()
+                isStreamCurrentlyAllowedToSendEvents = true
+            } catch {
+                throw DirectoryMonitorError.streamCouldNotBeStarted
+            }
+        }
+    }
+
+    /// Stops the monitor, preventing any further invocation of the event-
+    /// handler block.
+    func stop() {
+        // Stop sending events.
+        if isStreamCurrentlyAllowedToSendEvents {
+            stream?.stop()
+            isStreamCurrentlyAllowedToSendEvents = false
+        }
+
+        // Unschedule the stream from its run loop.
+        if isStreamCurrentlyScheduled {
+            stream?.invalidate()
+            isStreamCurrentlyScheduled = false
+        }
+
+        // Remove the (only) reference count before unsetting the pointer.
+        stream?.release()
+        stream = nil
+
+        // Release the event handler.
+        eventHandler = nil
+    }
+
+}
+
+// MARK: - Error handling
+
+enum DirectoryMonitorError: LocalizedError {
+    case noDirectoriesProvided
+    case directoryCannotBeOpened(path: String)
+    case streamCouldNotBeCreated
+    case streamCouldNotBeStarted
+
+    var errorDescription: String? {
+        switch self {
+        case .noDirectoriesProvided:
+            return "No directories have been provided"
+        case .directoryCannotBeOpened(let path):
+            return "The directory cannot be opened: \(path)"
+        case .streamCouldNotBeCreated:
+            return "The file-system event stream could not be created"
+        case .streamCouldNotBeStarted:
+            return "The file-system event stream could not be started"
+        }
+    }
+}
+
+// MARK: - Convenience extensions
+
+private typealias FileSystemEventStream = FSEventStreamRef
```

Comment:
Apple's docs say Swift's FSEventStreamRef is 10.10+, I'm sure that's not the cause of the test failing but it is something  we need to consider!